tee-worker/litentry/core/vc-task/receiver/src/lib.rs

```diff
@@ -84,62 +91,300 @@ pub fn run_vc_handler_runner<ShieldingKeyRepository, A, S, H, O, Z, N>(
+		) = tcs.call
+		{
+			// Filter out duplicate assertions
+			let mut seen: HashSet<H256> = HashSet::new();
+			let mut unique_assertions = Vec::new();
+			for assertion in assertions.into_iter() {
+				let hash = H256::from(blake2_256(&assertion.encode()));
+				if seen.insert(hash) {
+					unique_assertions.push(Some(assertion));
+				} else {
+					unique_assertions.push(None);
+				}
 			}
-			if response.is_err() {
-				if let Err(e) =
-					context_pool.ocall_api.update_metric(EnclaveMetric::FailedVCIssuance)
-				{
-					warn!("Failed to update metric for VC Issuance: {:?}", e);
+
+			let assertion_len = unique_assertions.len() as u8;
+			req_registry.add_new_item(connection_hash, assertion_len);
+			for (idx, assertion) in unique_assertions.iter().enumerate() {
+				if let Some(assertion) = assertion {
+					let new_call = TrustedCall::request_vc(
+						signer.clone(),
+						who.clone(),
+						assertion.clone(),
+						maybe_key,
+						req_ext_hash,
+					);
+
+					let shard_pool = request.shard;
+					let context_pool = context.clone();
+					let extrinsic_factory_pool = extrinsic_factory.clone();
+					let node_metadata_repo_pool = node_metadata_repo.clone();
+					let tc_sender_pool = tc_sender.clone();
+					let req_registry_pool = req_registry.clone();
+
+					pool.spawn_ok(async move {
+						let response = process_single_request(
+							shard_pool,
+							context_pool.clone(),
+							extrinsic_factory_pool,
+							node_metadata_repo_pool,
+							tc_sender_pool,
+							new_call,
+						);
+
+						//Totally fine to `unwrap` here. Because new item just added above.
+						let (size, len) =
+							req_registry_pool.update_item(connection_hash, idx as u8).unwrap();
+						send_vc_response(
+							connection_hash,
+							context_pool,
+							response,
+							idx as u8,
+							assertion_len,
+							size < len,
+						);
+					});
+				} else {
+					//Totally fine to `unwrap` here. Because new item just added above.
+					let (size, len) = req_registry.update_item(connection_hash, idx as u8).unwrap();
+					send_vc_response(
+						connection_hash,
+						context.clone(),
+						Err("Duplicate assertion request".to_string()),
+						idx as u8,
+						assertion_len,
+						size < len,
+					);
 				}
-			} else if let Err(e) =
-				context_pool.ocall_api.update_metric(EnclaveMetric::SuccessfullVCIssuance)
-			{
-				warn!("Failed to update metric for VC Issuance: {:?}", e);
 			}
-		});
+		} else {
+			send_vc_response(
+				connection_hash,
+				context.clone(),
+				Err("Wrong trusted call. Expect request_batch_vc ".to_string()),
+				0u8,
+				0u8,
+				false,
+			);
+		}
 	}
 	warn!("vc_task_receiver loop terminated");
 }
 
-pub fn handle_request<ShieldingKeyRepository, A, S, H, O, Z, N>(
-	request: &mut AesRequest,
+#[derive(Clone)]
+struct RequestRegistry {
+	status_map: Arc<Mutex<HashMap<H256, AssertionStatus>>>,
+}
+
+struct AssertionStatus {
+	pub time_stamp: Instant,
+	pub len: u8,
+	pub indexes: Vec<u8>,
```

Comment:
We could have better names, like `total` / `processed` etc

And it seems we don't need concrete indexes, just `processed_count`, do we?

Reply:
>  better names
Yes.
> concrete indexes
Originally planned to check idx and find any missing items. Looks like not necessary now.